### PR TITLE
Puns TMUse me

### PR DESCRIPTION
**Pull request by oleglazari**
Updated the current method for reading using the MAX31855 device on the TMU.
- Currently, the code does not implement any checking to see if the MAX31855 is returning any errors. 
- The updated code creates and raises certain types of errors using an error enum.
- Returns int max if any data is considered invalid due to an error.

### CODE REVIEW

File: include/TMU.hpp [hunk 1]
```diff
@@ -40,9 +40,10 @@ class TMU {
     uint16_t getObjectDictionarySize() const;
 
     /**
-     * Updates the thermTemps values
+     * Updates the temperature values in an array and updates the error array from the TMU object.
+     *
```
Comment (mjmagee991) [inline include/TMU.hpp]:
```suggestion
```

File: include/TMU.hpp [hunk 2]
```diff
@@ -58,7 +59,7 @@ class TMU {
     /**
```
Comment (mjmagee991) [inline include/TMU.hpp:63]: You need to declare the error array as a private variable here, so the object dictionary can reference it and output over CAN

File: include/TMU.hpp [hunk 3]
```diff
@@ -185,7 +187,79 @@ class TMU {
...
+        {
+            .Key = CO_KEY(0x1A01, 4, CO_UNSIGNED32 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = CO_LINK(0x2101, 3, 8),
+        },
+
+        /*
          * Data Links
          */
         {
```
Comment (mjmagee991) [inline include/TMU.hpp:269]: You need to add pointers to the error array here, so the data can be reported over CAN

File: include/dev/MAX31855.hpp [hunk 4]
```diff
@@ -3,6 +3,7 @@
 
 #include <EVT/io/SPI.hpp>
 #include <cstdint>
+#include <utility>
```
Comment (mjmagee991) [inline include/dev/MAX31855.hpp]: What's this for?

File: include/dev/MAX31855.hpp [hunk 5]
```diff
@@ -22,12 +23,31 @@ class MAX31855 {
...
+    /*
+     * The status of the MAX31855 after starting a temperature read.
+     *
+     * NO_ERROR: No error
+     * SCV_FAULT: Short to VCC
+     * SCG_FAULT: Short to GND
+     * OC_FAULT: Open Circuit
+     * MAX31855_ERROR: Error reading from the MAX31855
+     */
+    enum class MaxStatus {
```
Comment (mjmagee991) [inline include/dev/MAX31855.hpp]:
```suggestion
    enum MaxStatus {
```

File: include/dev/MAX31855.hpp [hunk 6]
```diff
@@ -22,12 +23,31 @@ class MAX31855 {
...
+        NO_ERROR = 0,
```
Comment (mjmagee991) [inline include/dev/MAX31855.hpp]: This is how we normally handle this
```suggestion
        OK = 0,
```

File: include/dev/MAX31855.hpp [hunk 7]
```diff
@@ -22,12 +23,31 @@ class MAX31855 {
...
+        SCV_FAULT = 1,
+        SCG_FAULT = 2,
+        OC_FAULT = 3,
+        MAX31855_ERROR = 4
```
Comment (mjmagee991) [inline include/dev/MAX31855.hpp]: Rename this flag to make it clearer

File: src/dev/MAX31855.cpp [hunk 8]
```diff
@@ -1,27 +1,69 @@
...
+MAX31855::MaxStatus MAX31855::readTemp(uint16_t& temp) {
+    uint8_t lastDataBit = 0;
+    uint16_t returned_temp = 0;
 
     const uint8_t length = 4;
-    uint8_t bytes[length];
+    uint8_t bytes[length] = {'\0'};
 
     spi.startTransmission(device);
     spi.read(bytes, length);
```
Comment (mjmagee991) [inline src/dev/MAX31855.cpp]: You should check here if the read fails and report a `SPI_ERROR` if it does

File: include/dev/MAX31855.hpp [hunk 9]
```diff
@@ -22,12 +23,31 @@ class MAX31855 {
...
+    };
+
     /**
-     * Returns the temp in centicelsius
+     * Reads the temperature from the MAX31855 and updates the temp variable passed in.
+     *
+     * @param[in] temp The temperature to update using the readTemp method.
```
Comment (mjmagee991) [inline include/dev/MAX31855.hpp]:
```suggestion
     * @param[out] temp The temperature reported by the IC
```

File: src/TMU.cpp [hunk 10]
```diff
@@ -13,9 +13,17 @@ uint16_t TMU::getObjectDictionarySize() const {
     return OBJECT_DICTIONARY_SIZE;
 }
 
-void TMU::updateTemps() {
+void TMU::process() {
+    uint16_t temp_array[4] = {0};
```
Comment (mjmagee991) [inline src/TMU.cpp]: Using this array instead of `thermTemps` prevents the data from being read by CANopen properly. You should go back to using that array

File: src/dev/MAX31855.cpp [hunk 11]
```diff
@@ -1,27 +1,69 @@
...
 
-uint16_t MAX31855::readTemp() {
-    uint16_t temp;
+MAX31855::MaxStatus MAX31855::readTemp(uint16_t& temp) {
+    uint8_t lastDataBit = 0;
+    uint16_t returned_temp = 0;
 
     const uint8_t length = 4;
-    uint8_t bytes[length];
+    uint8_t bytes[length] = {'\0'};
```
Comment (mjmagee991) [inline src/dev/MAX31855.cpp]: Why did you add this?

File: src/dev/MAX31855.cpp [hunk 12]
```diff
@@ -1,27 +1,69 @@
...
 using namespace EVT::core::IO;
 
 namespace TMU::DEV {
 
 MAX31855::MAX31855(SPI& spi, uint8_t device) : spi(spi), device(device) {}
 
-uint16_t MAX31855::readTemp() {
-    uint16_t temp;
+MAX31855::MaxStatus MAX31855::readTemp(uint16_t& temp) {
+    uint8_t lastDataBit = 0;
```
Comment (mjmagee991) [inline src/dev/MAX31855.cpp]: Why do you declare this up here?

File: src/dev/MAX31855.cpp [hunk 13]
```diff
@@ -1,27 +1,69 @@
...
     spi.endTransmission(device);
 
-    temp = (((uint16_t) bytes[0]) << 8) | bytes[1];
+    returned_temp = (((uint16_t) bytes[0]) << 8) | bytes[1];
 
-    temp = temp >> 2;                             // lose last 2 bits as they aren't in the 14 bit temp
-    temp = (temp >> 2) * 100 + (temp & 0x03) * 25;// convert last 2 digits to allow for decimal place
-    return temp;
+    lastDataBit = ((bytes[1]) << 7);// Shift it 7 to get rid of internal temp data and reserve bit
+    lastDataBit = lastDataBit >> 7; // Shift it back to the least significant position
```
Comment (mjmagee991) [inline src/dev/MAX31855.cpp]: You can do this more easily with `lastDataBit = bytes[1] & 0x01`

File: src/dev/MAX31855.cpp [hunk 14]
```diff
@@ -1,27 +1,69 @@
...
+     * 0x02 is the flag for SCG Fault (Short to GND)
+     * 0x04 is the flag for SCV Fault (Short to VCC)
+
+     Check the error checking bit, if bit equals 1, then there is an error
+     */
+    if (lastDataBit & 0x01) {
+        uint8_t lastByte = 0;
+        lastByte = bytes[3];
+        lastByte = lastByte << 5;//Shift it 5 to get rid of internal temp data and reserve bit
+        lastByte = lastByte >> 5;//Shift it back to the least significant position
```
Comment (mjmagee991) [inline src/dev/MAX31855.cpp]: You can use `&` to simplify this too

File: src/dev/MAX31855.cpp [hunk 15]
```diff
@@ -1,27 +1,69 @@
...
+
+        // Check for OC Fault
+        if (lastByte & 0x01) {
+
```
Comment (mjmagee991) [inline src/dev/MAX31855.cpp]:
```suggestion
```

File: src/dev/MAX31855.cpp [hunk 16]
```diff
@@ -1,27 +1,69 @@
...
 
 namespace TMU::DEV {
 
 MAX31855::MAX31855(SPI& spi, uint8_t device) : spi(spi), device(device) {}
 
-uint16_t MAX31855::readTemp() {
-    uint16_t temp;
+MAX31855::MaxStatus MAX31855::readTemp(uint16_t& temp) {
+    uint8_t lastDataBit = 0;
+    uint16_t returned_temp = 0;
```
Comment (mjmagee991) [inline src/dev/MAX31855.cpp]: This name makes it a bit clearer that the data is temperature and more
```suggestion
    uint16_t returned_data = 0;
```

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
Great start! Just a few things to clean up

File: include/dev/MAX31855.hpp (hunk 4 above)
Comment (oleglazari) [reply include/dev/MAX31855.hpp]: pretty sure I had this when I was trying to implement pair, its some useless code.

File: src/dev/MAX31855.cpp (hunk 11 above)
Comment (oleglazari) [reply src/dev/MAX31855.cpp]: I just don't want any crummy data when the array gets initialized, its just a habit of mine.

File: src/dev/MAX31855.cpp [hunk 17]
```diff
@@ -7,21 +7,61 @@ namespace TMU::DEV {
...
     spi.startTransmission(device);
-    spi.read(bytes, length);
+    SPI::SPIStatus status = spi.read(bytes, length);
+    if (status != SPI::SPIStatus::OK) {
+        return MaxStatus::SPI_ERROR;
+    }
     spi.endTransmission(device);
 
-    temp = (((uint16_t) bytes[0]) << 8) | bytes[1];
+    returned_data = (((uint16_t) bytes[0]) << 8) | bytes[1];
```
Comment (ActuallyTaylor) [inline src/dev/MAX31855.cpp]: This declaration should probably be moved down to where the other `returned_data` calculations are happening. I think the variable should also be renamed to `returnedData` to match the camel case from the rest of the project.

File: targets/MAX31855/main.cpp [hunk 18]
```diff
@@ -38,11 +38,17 @@ int main() {
         TMU::DEV::MAX31855(spi, 3),
     };
 
-    uart.printf("read start:\r\n");
+    uart.printf("read start:\r\n");// Print a message to the UART
```
Comment (ActuallyTaylor) [inline targets/MAX31855/main.cpp]:
```suggestion
    uart.printf("read start:\r\n");// Print a message to UART
```

File: include/TMU.hpp [hunk 19]
```diff
@@ -185,7 +191,79 @@ class TMU {
...
+        },
+        {
+            //no inhibit time
+            .Key = CO_KEY(0x1801, 3, CO_UNSIGNED16 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = (uintptr_t) 0,
+        },
+        {
+            //send every 2 seconds
+            .Key = CO_KEY(0x1801, 4, CO_UNSIGNED16 | CO_OBJ_D__R_),
```
Comment (ActuallyTaylor) [inline include/TMU.hpp]:
```suggestion
            .Key = CO_KEY(0x1801, 5, CO_UNSIGNED16 | CO_OBJ_D__R_),
```
A value of 4 here is the wrong key for CANOpen, same as above

File: include/TMU.hpp [hunk 20]
```diff
@@ -185,7 +191,79 @@ class TMU {
...
+        {
+            .Key = CO_KEY(0x1801, 0, CO_UNSIGNED8 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = (uintptr_t) 5,
+        },
+        {
+            //180h+Node-ID
+            .Key = CO_KEY(0x1801, 1, CO_UNSIGNED32 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = (uintptr_t) CO_COBID_TPDO_DEFAULT(0) + NODE_ID,
```
Comment (ActuallyTaylor) [inline include/TMU.hpp]:
```suggestion
            .Data = (uintptr_t) CO_COBID_TPDO_DEFAULT(1) + NODE_ID,
```

This needs to be advanced to `CO_COBID_TPDO_DEFAULT(1)`. If it is set as 0 it will either confused CANOpen or send to PDOs with the same ID

File: include/TMU.hpp [hunk 21]
```diff
@@ -147,12 +152,13 @@ class TMU {
         },
         {
             //send every 2 seconds
-            .Key = CO_KEY(0x1800, 5, CO_UNSIGNED16 | CO_OBJ_D__R_),
+            .Key = CO_KEY(0x1800, 4, CO_UNSIGNED16 | CO_OBJ_D__R_),
```
Comment (ActuallyTaylor) [inline include/TMU.hpp]:
```suggestion
            .Key = CO_KEY(0x1800, 5, CO_UNSIGNED16 | CO_OBJ_D__R_),
```

This should still be 5. Changing it to 4 overlaps with a reserved value that you are not using. This is mostly an issue with how CANOpen dictionaries are created...

**Review by ActuallyTaylor — CHANGES_REQUESTED** [review]
There are some issues with your CANOpen dictionary. Some keys that you changed will result in the dictionary not properly functioning. Other than that, the error handling looks solid.

File: include/TMU.hpp [hunk 22]
```diff
@@ -208,6 +286,30 @@ class TMU {
             .Type = nullptr,
             .Data = (uintptr_t) &thermTemps[3],
         },
+
+        /*
+         * Error Data Links
+         */
+        {
+            .Key = CO_KEY(0x2101, 0, CO_UNSIGNED8 | CO_OBJ___PRW),
```
Comment (mjmagee991) [inline include/TMU.hpp:294]: When adding more data links, you only increment the subindex, not the main index. This goes for the other error array data as well
```suggestion
            .Key = CO_KEY(0x2100, 4, CO_UNSIGNED8 | CO_OBJ___PRW),
```
Make sure you update the Mapping links as well

File: include/TMU.hpp [hunk 23]
```diff
@@ -185,7 +191,79 @@ class TMU {
             .Data = CO_LINK(0x2100, 3, 16),
         },
 
-        /**
+        /*
```
Comment (mjmagee991) [inline include/TMU.hpp]: Normally, we put all the settings sections and then all the mapping sections, and then all the data. There's no particular reason for this, but we do want to be consistent across repositories, so it's easy for people to understand and work on multiple projects. Here, all this means is that I'd like you to move this TPDO 1 Settings section up above TPDO 0 Mapping

File: include/dev/MAX31855.hpp (hunk 4 above)
Comment (mjmagee991) [reply include/dev/MAX31855.hpp]: If it's useless, can you delete it?

File: include/dev/MAX31855.hpp [hunk 24]
```diff
@@ -22,12 +23,32 @@ class MAX31855 {
      */
     explicit MAX31855(SPI& spi, uint8_t device);
 
+    /*
```
Comment (mjmagee991) [inline include/dev/MAX31855.hpp]: We want this to be picked up by Doxygen, so it should have two asterisks. Can you see if there's a standard way of documenting enums in Doxygen?
```suggestion
    /**
```

File: include/dev/MAX31855.hpp (hunk 9 above)
Comment (mjmagee991) [reply include/dev/MAX31855.hpp]: You added the proper documentation, but you didn't delete the improper stuff

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
A few more details to iron out

File: src/TMU.cpp [hunk 25]
```diff
@@ -13,9 +13,16 @@ uint16_t TMU::getObjectDictionarySize() const {
     return OBJECT_DICTIONARY_SIZE;
 }
 
-void TMU::updateTemps() {
+void TMU::process() {
+    uint16_t thermTemps[NUM_THERMOCOUPLES] = {};
+
     for (int i = 0; i < 4; i++) {
```
Comment (mjmagee991) [inline src/TMU.cpp]: This doesn't make a huge difference, but it is a bit more consistent this way
```suggestion
    for (int i = 0; i < NUM_THERMOCOUPLES; i++) {
```

**Review by ActuallyTaylor — APPROVED** [review]
The stuff I had problems with is fixed. Besides what Matt said about the CAN Dictionary ordering this looks good.

File: include/TMU.hpp [hunk 26]
```diff
@@ -185,7 +229,41 @@ class TMU {
...
+         */
+        {
+            .Key = CO_KEY(0x1A00, 5, CO_UNSIGNED8 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = (uintptr_t) 4,
+        },
+        {
+            .Key = CO_KEY(0x1A00, 6, CO_UNSIGNED32 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = CO_LINK(0x2101, 0, 8),
```
Comment (mjmagee991) [inline include/TMU.hpp:248]: Since you fixed the section below, you need to fix this as well
```suggestion
            .Data = CO_LINK(0x2100, 0, 8),
```

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
Double-check your object dictionary

File: include/TMU.hpp (hunk 26 above)
Comment (mjmagee991) [reply include/TMU.hpp:248]: There are more issues with this in the lines following this. Also, please validate that the data is being reported properly over CAN before you merge this

**Comment by oleglazari** [comment]
everything fully work with can ඞ

File: targets/DEV1-TMU/main.cpp [hunk 27]
```diff
@@ -76,7 +76,15 @@ extern "C" void CONmtHbConsChange(CO_NMT* nmt, uint8_t nodeId, CO_MODE mode) {}
 
 extern "C" int16_t COParaDefault(CO_PARA* pg) { return 0; }
 
-extern "C" void COPdoTransmit(CO_IF_FRM* frm) {}
+extern "C" void COPdoTransmit(CO_IF_FRM* frm) {
+    log::LOGGER.log(log::Logger::LogLevel::DEBUG, "Sending PDO as 0x%X with length %d and data: ", frm->Identifier, frm->DLC);
```
Comment (mjmagee991) [inline targets/DEV1-TMU/main.cpp]: I'm sure this was useful for debugging, but can you revert this back now that you've confirmed things are working?

**Review by mjmagee991 — APPROVED** [review]
One thing to clean up. Otherwise, great job!
